src/shaka_scheme/system/gc/mark_procedures.cpp

```diff
@@ -0,0 +1,76 @@
+//
+// Created by mortrax on 4/16/18.
+//
+#include "shaka_scheme/system/gc/GCNode.hpp"
+#include "shaka_scheme/system/vm/HeapVirtualMachine.hpp"
+#include "shaka_scheme/system/gc/GCData.hpp"
+
+namespace shaka {
+namespace gc {
+
+static void mark_accumulator(Accumulator a) {
+}
+
+static void mark_expression(Expression e) {
+}
+
+static void mark_environment(Environment env) {
+}
+
+static void mark_call_frame(CallFrame f) {
+}
+
+static void mark_value_rib(ValueRib vr) {
+}
+
+void mark(const HeapVirtualMachine& hvm) {
+  mark_accumulator(hvm.get_accumulator());
+  mark_expression(hvm.get_expression());
+  mark_environment(*hvm.get_environment());
+  mark_call_frame(*hvm.get_call_frame());
+  mark_value_rib(hvm.get_value_rib());
+}
+
+void mark_node(const GCNode& node) {
+
+  // Check for cycles
+  if (node.gc_data->is_marked()) {
+    return;
+  }
+
+  node.gc_data->mark();
+  switch(node->get_type()) {
+  case shaka::Data::Type::DATA_PAIR: {
+    // If the argument is a pair, mark its car and cdr
+    mark_node(node->get<DataPair>().car());
+    mark_node(node->get<DataPair>().cdr());
```

Comment:
Indeed, this was discussed. Here is my rationale. In the case of most Scheme lists, we are dealing with linear chains, meaning that the maximum depth of recursion should be O(n), linear in the number of cons cells in the list on average. Now, there are certainly exceptions to this rule, as you can construct arbitrarily complex data structures using scheme pairs, with long chains of cons  cells attached at both the `car` and `cdr` position of a given node. With that being said, a recursive implementation like this, which is a mere two lines of code, while potentially expensive in terms of stack space usage, is much simpler than any sufficient iterative approach. When you consider the fact that you can have arbitrarily nested `DataPair`s, potentially with `Vector`s inside of them, that may have other `DataPair`s inside of them, which may have `Closure`s inside of them, etc, etc. this high level of dynamism in the way that data structures can be linked in the system lends itself well to a recursive approach. If it turns out later that this approach is not feasible when used in production, then we can look at optimizing parts of the algorithm by slotting in iterative chunks as needed. But for right now, I would prefer to stick to this approach, as the code is much simpler to understand, and the constraints of the system too complex for me to reason about up front.

Reply:
Another thing I will say is, while the recursion tree could potentially be complex for this algorithm, the sub-problems solved by the algorithm will all be disjoint. Why? Because the if check on line 37 of `mark_node()` ensures that we only process nodes that we haven't already marked, acting as a sort of memoization that prevents us from ever doing more work than is needed. Again, we can always optimize parts of the algorithm later if we find that we are blowing the stack during the mark phase of execution, but for now, I would like to stick with the simpler implementation that is known to work, and not optimize prematurely until we have seen how the system performs in production.